Lib/fontbakery/profiles/universal.py

```diff
@@ -2212,6 +2213,189 @@ def com_google_fonts_check_math_signs_width(ttFont):
+        return ttFont.getReverseGlyphMap()[glyph_name]
+
+    def unicode_for_glyph(ttFont, glyph_name):
+        for table in ttFont["cmap"].tables:
+            if table.isUnicode():
+                for codepoint, glyph in table.cmap.items():
+                    if glyph == glyph_name:
+                        return codepoint
+
+    def has_feature(ttFont, featureTag):
+        if "GSUB" in ttFont and ttFont["GSUB"].table.FeatureList:
+            for FeatureRecord in ttFont["GSUB"].table.FeatureList.FeatureRecord:
+                if FeatureRecord.FeatureTag == featureTag:
+                    return True
+        if "GPOS" in ttFont and ttFont["GPOS"].table.FeatureList:
+            for FeatureRecord in ttFont["GPOS"].table.FeatureList.FeatureRecord:
+                if FeatureRecord.FeatureTag == featureTag:
+                    return True
+        return False
+
+    def buf_to_width(buf):
+        x_cursor = 0
+
+        for _info, pos in zip(buf.glyph_infos, buf.glyph_positions):
+            x_cursor += pos.x_advance
+
+        return x_cursor
+
+    def get_substitutions(
+        ttFont,
+        featureTag,
+        lookupType=1,
+    ):
+        substitutions = {}
+        if "GSUB" in ttFont:
+            for FeatureRecord in ttFont["GSUB"].table.FeatureList.FeatureRecord:
+                if FeatureRecord.FeatureTag == featureTag:
+                    for LookupIndex in FeatureRecord.Feature.LookupListIndex:
+                        for SubTable in (
+                            ttFont["GSUB"].table.LookupList.Lookup[LookupIndex].SubTable
+                        ):
+                            if SubTable.LookupType == lookupType:
+                                substitutions.update(SubTable.mapping)
+        return substitutions
+
+    def nominal_glyph_func(font, code_point, data):
+        return code_point
+
+    def buffer_for_GIDs(GIDs, features):
+        buf = hb.Buffer()
+        buf.add_codepoints(GIDs)
+        buf.guess_segment_properties()
+        funcs = hb.FontFuncs.create()
+        funcs.set_nominal_glyph_func(nominal_glyph_func)
+        vhb.hbfont.funcs = funcs
+        hb.shape(vhb.hbfont, buf, features)
+        return buf
+
+    def kerning(ttFont, glyph_list):
+        GID_list = [GID_for_glyph(ttFont, glyph) for glyph in glyph_list]
+        return buf_to_width(
+            buffer_for_GIDs(
+                GID_list,
+                {"kern": True},
+            )
+        ) - buf_to_width(
+            buffer_for_GIDs(
+                GID_list,
+                {"kern": False},
+            )
+        )
+
+    all_glyphs = list(ttFont.getGlyphOrder())
+    tabular_glyphs = []
+    tabular_numerals = []
+
+    # Fonts with tnum feautre
+    if has_feature(ttFont, "tnum"):
+        tabular_glyphs = list(get_substitutions(ttFont, "tnum").values())
+        buf = vhb.shape("0123456789", {"features": {"tnum": True}})
+        tabular_numerals = vhb.serialize_buf(buf, glyphsonly=True).split("|")
```

Comment:
If you are accessing the GSUB table directly to get `tabular_glyphs`, why do you need HarfBuzz to get `tabular_numerals` instead of getting the mapping from `tabular_glyphs`?

Reply:
For the odd case where encoded numbers aren't the input characters to the tnum feature, but are first shaped into other glyphs and later shaped into tabular numerals. I'm not saying this is the norm, but I wanted to be thorough here.